README.md
# eth-zymbit-keyring

An implementation of MetaMask's [Keyring interface](https://github.com/MetaMask/eth-simple-keyring#the-keyring-class-protocol) for Zymbit devices with built in hardware wallets ([HSM6](https://www.zymbit.com/hsm6/) and [SCM](https://www.zymbit.com/scm/)).

For the most part, it works the same way as [eth-hd-keyring](https://github.com/MetaMask/eth-hd-keyring), but uses a Zymbit HSM6 or SCM to store the keys and perform any signing operations. However, there are some differences:

- It requires that you have an HSM6 or SCM attached to a supported embedded device (Raspberry Pi or Jetson Nano)
- All cryptographic operations involving keys are performed on the module itself
- Zymbit devices don't support exporting private keys, so it does not support the `exportAccount()` method 

Installation
-------
Run the following command:

`yarn add eth-zymbit-keyring`

or

`npm install eth-zymbit-keyring`

Usage
-------

```
const ZymbitKeyring = require('eth-zymbit-keyring')
const options = {
    wallet_name: "MyExampleWallet",
    master_slot: 16
}
const keyring = new ZymbitKeyring(options);
```

### Keyring.type

A class property that returns `Zymbit Hardware Wallet`

### constructor( options )

The constructor receives an options object that requires either a valid `wallet_name` or `master_slot` property. In order to use this, you must have created or recovered a BIP39 wallet through [Zymbit's API](https://docs.zymbit.com/tutorials/digital-wallet/wallet-example/).

## Zymbit Keyring Instance Methods

All below instance methods return Promises to allow asynchronous resolution.

### serialize()

Retuerns a JSON-serializable object with the `wallet_name` or `master_slot` properties of the keyring. This object can be passed into `deserialize()` to recover the state of the keyring. 

### deserialize( object )

As discussed above, the `deserialize()` method will be passed the JSON object that was returned when the `serialize()` method was called.

### addAccounts( n = 1 )

Used to add new accounts to the keyring starting from `m/44'/60'/0'/0` and incrementing the last index each time. Takes the number of accounts you want to add to the keyring `n` as a parameter (defaults to `n = 1`), and returns an array of size `n` including the Ethereum addresses of the accounts added.

### getAccounts()

Returns an array of hex-string addresses for the Ethereum accounts that the keyring is able to sign for.

### signTransaction(address, transaction)

Takes a hex-prefixed `address` string for the account you want to use to sign the incoming `transaction` with. The `transaction` is an instance of [ethereumjs-tx](https://github.com/ethereumjs/ethereumjs-tx). Returns a signed [ethereumjs-tx](https://github.com/ethereumjs/ethereumjs-tx) object when complete.

### signMessage(address, data)

Takes a pre-hashed message (`data`) and signs it using the account of the hex-prefixed `address` passed in. Returns the complete hex-prefixed ECDSA signature.

### removeAccount(address)

Removes the account corresponding to the hex-prefixed `address` passed in from the keyring and the Zymbit device.

Testing
-------
Run the following command:

`npm test`

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Added
- changelong
- npm
- npmignore
- update changelog
- publish-script
- reset changelog
- unreleased
- release
- readme
- changelog
- fixChangeLog
- fix
- workflow
- changelong
- npm
- readme
- completed initial tests
- testingDone
- signMessage test works
- signed tx test done
- more tests done
- more tests done
- more tests done
- setup testing with mocha
- signing messages works
- transaction signature works
- signKindOfWorks
- formatting
- exportAccounts
- Revert "getAccounts works"
- Revert "exportAccount"
- exportAccount
- getAccounts works
- addAccounts works
- constructor and deserialize
- built js wrapper for zk
- zkAppUtilsClass.h added
- Merge branch 'main' of gitlab.com:zymbit/eth-zymbit-keyring
- setup class
- readme
- Merge branch 'main' of https://gitlab.com/zymbit/eth-zymbit-keyring
- setup
- Initial commit


[Unreleased]: git+https://github.com/ShivaanshK/eth-zymbit-keyring/


test.py
import zymkey 

packet =(bytes. ("b2c0a7451cf0bd1bdc86ee3fbbbfe6baf502670bae2a71cd093b37f6ff4c4102"))
print(packet)
   
signature, y_parity = zymkey.client.sign(
    packet, slot=22, return_recid=True)
N = 115792089237316195423570985008687907852837564279074904382605163141518161494337
r = int.from_bytes(signature[:32], "big")
s = int.from_bytes(signature[-32:], "big")

y_parity = bool(y_parity.value)
if((s*2) >= N):
    y_parity = not y_parity
    s = N - s

v = 1 * 2 + 35 + int(y_parity)

signature = '0x'+hex(r)[2:].zfill(64)+hex(s)[2:].zfill(64)+hex(v)[2:]

print(signature)

src/test.js
const keyring = require('./eth-zymbit-keyring')
const zkJS = require('../zkJS/build/Release/zkAppUtilsJS.node');
const { bytesToHex, bytesToUtf8 } = require('ethereum-cryptography/utils');
const { type } = require('./eth-zymbit-keyring');
const { ethers } = require("ethers");
const EthereumTx = require('ethereumjs-tx').Transaction
const {
  BN,
  bufferToInt,
  ecrecover,
  rlphash,
  publicToAddress,
  ecsign,
  toBuffer,
  rlp,
  stripZeros,
} =  require('ethereumjs-util')

const zk = new zkJS.zkObj()
opts = {
    wallet_name: "MyExampleWallet",
    master_slot: 16
}

// for (const value of inputBuffer.values()) {
//     console.log(value);
//   }
// const intArray = Array.from({ length: inputBuffer.length / 5 }).map((_, i) => inputBuffer.readUInt32BE(i * 5));


const zymbitKeyring = new keyring(opts)

zymbitKeyring.addAccounts()

let accounts = zymbitKeyring.getAccounts()

let fromAccount = accounts[0]

const txParams = {
  from: fromAccount,
  nonce: '0x00',
  gasPrice: '0x09184e72a000',
  gasLimit: '0x2710',
  to: accounts[2],
  value: '0x1000',
};
  
  // The second parameter is not necessary if these values are used
  let tx = new EthereumTx(txParams, { chain: 'mainnet'})

const message =
  '0x879a053d4800c6354e76c7985a865d2922c82fb5b3f4577b2fe08b998954f2e0';

tx = zymbitKeyring.signTransaction(fromAccount.toLowerCase(), tx)
console.log(tx.verifySignature())

console.log(zymbitKeyring.signMessage(fromAccount, message))

// console.log(zymbitKeyring.addAccounts(3))


tests/test-eth-zymbit-keyring.js
const assert = require('assert')
const zkJS = require('../zkJS/build/Release/zkAppUtilsJS.node');

const ZymbitKeyring = require('../src/eth-zymbit-keyring');
const { bytesToHex } = require('ethereum-cryptography/utils');
const { ethers } = require('ethers');
const { randomInt } = require('crypto');
const zk = new zkJS.zkObj()
const EthereumTx = require('ethereumjs-tx').Transaction
const { stripHexPrefix } = require('ethereumjs-util')
const { toBuffer, ecrecover, pubToAddress } = require('@ethereumjs/util');

const getStartingSlots = () => {
    const startingSlots = []
    const slotsBuffer = zk.getAllocSlotsList(false)
    for (let i = 0; i < slotsBuffer.length / 4; i++) {
        const slot = slotsBuffer.readInt32LE(4 * i)
        if (slot > 15) startingSlots.push(slot)
    }
    return startingSlots
}

const removeKeysCreated = startingSlots => {
    const slotsBuffer = zk.getAllocSlotsList(false)
    for (let i = 0; i < slotsBuffer.length / 4; i++) {
        const slot = slotsBuffer.readInt32LE(4 * i)
        if (slot > 15 && !startingSlots.includes(slot)) zk.removeKey(slot, false)
    }
}


describe('ZymbitKeyring', function () {
    let startingSlots, wallet_name, master_slot, opts, keyring

    before('Setup Keyring', function (done) {
        this.timeout(20000)
        startingSlots = getStartingSlots()
        wallet_name = "TestWallet"
        const { slot, mnemonic } = zk.genWalletMasterSeedWithBIP39("secp256k1", wallet_name, "", "")
        master_slot = slot
        opts = {
            wallet_name,
            master_slot: slot
        }
        keyring = new ZymbitKeyring(opts)
        done()
    })

    describe('constructor', function () {

        let baseSlotDetails
        before('Get base_slot details', function () {
            baseSlotDetails = zk.getWalletNodeAddrFromKeySlot(keyring.base_slot)
        })
        it('constructs', function () {
            assert.equal(typeof keyring, 'object')
        })
        it('initialzes keyring correctly', function () {
            assert.equal(ZymbitKeyring.type, 'Zymbit Hardware Wallet')
            assert.equal(ZymbitKeyring.base_path, baseSlotDetails.node_address)
            assert.equal(keyring.wallet_name, wallet_name)
            assert.equal(keyring.master_slot, master_slot)
            assert.equal(keyring.wallet_name, baseSlotDetails.wallet_name)
            assert.deepEqual(keyring.account_slots, [])
        })
    })

    describe('addAccounts', function () {
        let numAccounts, accounts, publicKeys
        before('Add accounts', function (done) {
            this.timeout(20000)
            numAccounts = randomInt(3, 11)
            accounts = keyring.addAccounts(numAccounts)
            publicKeys = keyring.account_slots.map(account_slot => zk.exportPubKey(account_slot.slot, false))
            done()
        })

        it('returns correct number of addresses', function () {
            assert.equal(accounts.length, numAccounts)
        })

        it('returns valid ethereum addresses', function () {
            accounts.forEach(account => {
                assert.equal(ethers.isAddress(account), true)
            })
        })

        it('returns correct addresses', function () {
            accounts.forEach((account, index) => {
                assert.equal(account, ethers.computeAddress('0x' + bytesToHex(publicKeys[index])))
            })
        })
    })

    describe('getAccounts', function () {
        let accounts, publicKeys
        before('Get accounts', function (done) {
            this.timeout(20000)
            accounts = keyring.getAccounts()
            publicKeys = keyring.account_slots.map(account_slot => zk.exportPubKey(account_slot.slot, false))
            done()
        })

        it('returns correct number of accounts', function () {
            assert.equal(accounts.length, keyring.account_slots.length)
        })

        it('returns correct addresses', function () {
            accounts.forEach((account, index) => {
                assert.equal(account, ethers.computeAddress('0x' + bytesToHex(publicKeys[index])))
            })
        })
    })

    describe('signTransaction', function () {
        let accounts, signerIndex, signer, transaction
        before('Setup Transaction', function (done) {
            this.timeout(20000)
            accounts = keyring.getAccounts()
            signerIndex = randomInt(0, accounts.length)
            signer = accounts[signerIndex]
            const txParams = {
                from: signer,
                nonce: '0x00',
                gasPrice: '0x09184e72a000',
                gasLimit: '0x2710',
                to: signer,
                value: '0x1000',
            };

            transaction = new EthereumTx(txParams, { chain: 'mainnet' })
            done()
        })

        it('signs transaction with correct address', function (done) {
            this.timeout(20000)
            const signedTransaction = keyring.signTransaction(signer.toLowerCase(), transaction)
            assert.equal(signedTransaction.verifySignature(), true)
            done()
        })
    })

    describe('signMessage', function () {
        let accounts, signerIndex, signer, message
        before('Setup Transaction', function (done) {
            this.timeout(20000)
            accounts = keyring.getAccounts()
            signerIndex = randomInt(0, accounts.length)
            signer = accounts[signerIndex]
            message = '0x' + [...Array(64)].map(() => Math.floor(Math.random() * 16).toString(16)).join('');
            done()
        })

        it('signs message with correct address', function (done) {
            this.timeout(20000)
            const signature = keyring.signMessage(signer.toLowerCase(), message);
            const r = toBuffer(signature.slice(0, 66));
            const s = toBuffer(`0x${signature.slice(66, 130)}`);
            const v = BigInt(`0x${signature.slice(130, 132)}`);
            const m = toBuffer(message);
            const pub = ecrecover(m, v, r, s);
            const expectedAddress = ethers.getAddress(`0x${pubToAddress(pub).toString('hex')}`)
            assert.deepEqual(expectedAddress, signer)
            done()
        })
    })

    describe('serialize', function () {
        let serializedKeyring
        before('Get accounts', function () {
            serializedKeyring = keyring.serialize()
        })
        it('returns correct wallet_name and slot', function () {
            assert.equal(serializedKeyring.wallet_name, wallet_name)
            assert.equal(serializedKeyring.master_slot, master_slot)
        })
    })


    describe('deserialize', function () {
        let temp_wallet_name, temp_keyring
        before('Setup Temporary Keyring', function (done) {
            this.timeout(20000)
            temp_wallet_name = "TestWallet1"
            const { temp_slot, temp_mnemonic } = zk.genWalletMasterSeedWithBIP39("secp256k1", temp_wallet_name, "", "")
            temp_opts = {
                wallet_name: temp_wallet_name,
                master_slot: temp_slot
            }
            temp_keyring = new ZymbitKeyring(temp_opts)
            done()
        })

        it('restores keyring\'s state', function (done) {
            this.timeout(20000)
            const serializedKeyring = keyring.serialize()
            temp_keyring.deserialize(serializedKeyring)
            assert.deepEqual(keyring, temp_keyring)
            done()
        })
    })

    after('Cleanup after testing', function (done) {
        this.timeout(20000)
        removeKeysCreated(startingSlots)
        done()
    })

})



zkJS/src/zkAppUtilsClassJS.h
#include <napi.h>
#include "./zkAppUtilsClass.h"
#include <string>
#include <sstream>
#include <iostream>
#include <vector>

class zkObj : public Napi::ObjectWrap<zkObj>
{
 public:
  static Napi::Object Init(Napi::Env env, Napi::Object exports); //Init function for setting the export key to JS
  zkObj(const Napi::CallbackInfo& info); //Constructor to initialise

 private:
  static Napi::FunctionReference constructor; //reference to store the class definition that needs to be exported to JS
  Napi::Value getTime(const Napi::CallbackInfo& info);
  Napi::Value exportPubKey(const Napi::CallbackInfo& info);
  Napi::Value storeForeignPubKey(const Napi::CallbackInfo& info);
  Napi::Value genKeyPair(const Napi::CallbackInfo& info);
  Napi::Value genEphemeralKeyPair(const Napi::CallbackInfo& info);
  Napi::Value removeKey(const Napi::CallbackInfo& info);
  Napi::Value invalidateEphemeralKey(const Napi::CallbackInfo& info);
  Napi::Value genECDSASigFromDigest(const Napi::CallbackInfo& info);
  Napi::Value verifyECDSASigFromDigest(const Napi::CallbackInfo& info);
  Napi::Value genWalletMasterSeedWithBIP39(const Napi::CallbackInfo& info);
  Napi::Value genWalletMasterSeedWithSLIP39(const Napi::CallbackInfo& info);
  Napi::Value setGenSLIP39GroupInfo(const Napi::CallbackInfo& info);
  Napi::Value addGenSLIP39Member(const Napi::CallbackInfo& info);
  Napi::Value cancelSLIP39Session(const Napi::CallbackInfo& info);
  Napi::Value genOverSightWallet(const Napi::CallbackInfo& info);
  Napi::Value genWalletChildKey(const Napi::CallbackInfo& info);
  Napi::Value restoreWalletMasterSeedFromBIP39(const Napi::CallbackInfo& info);
  Napi::Value restoreWalletMasterSeedFromSLIP39(const Napi::CallbackInfo& info);
  Napi::Value addRestoreSLIP39Mnemonic(const Napi::CallbackInfo& info);
  Napi::Value getWalletNodeAddrFromKeySlot(const Napi::CallbackInfo& info);
  Napi::Value getWalletKeySlotFromNodeAddr(const Napi::CallbackInfo& info);
  Napi::Value getAllocSlotsList(const Napi::CallbackInfo& info);
  zkAppUtils::zkClass *_zkCTX; //internal instance of zk obj used to perform actual operations.
};


src/eth-zymbit-keyring.js
const zkJS = require('zymkey-js/build/Release/zkAppUtilsJS.node');
const { ethers } = require("ethers");
const { bytesToHex } = require('ethereum-cryptography/utils');
const { BN, stripHexPrefix } = require('ethereumjs-util')
const { concatSig } = require('@metamask/eth-sig-util');

const BIP44_ETH_BASE_PATH = `m/44'/60'/0'/0`
const type = 'Zymbit Hardware Wallet'
const zymkey = new zkJS.zkObj()

class ZymbitKeyring {

  constructor(opts = {}) {
    this.deserialize(opts);
  }

  serialize() {
    return Promise.resolve({
      wallet_name: this.wallet_name,
      master_slot: this.master_slot
    })
  }

  deserialize(opts = {}) {
    if (!opts.wallet_name && (typeof (opts.master_slot) != 'number' || opts.master_slot < 16 || opts.master_slot > 512)) {
      throw new Error("Valid wallet name or master slot required")
    }
    if (opts.wallet_name) {
      try {
        this.master_slot = zymkey.getWalletKeySlotFromNodeAddr("m", opts.wallet_name)
        this.wallet_name = opts.wallet_name
      } catch {
        throw new Error("Invalid Wallet Name")
      }
    } else {
      try {
        const slotDetails = zymkey.getWalletNodeAddrFromKeySlot(opts.master_slot)
        if (slotDetails.node_address == 'm') {
          this.master_slot = opts.master_slot
          this.wallet_name = slotDetails.wallet_name
        } else {
          throw new Error
        }
      } catch (e) {
        throw new Error("Invalid Master Key Slot")
      }
    }
    this.base_slot = 0
    this.account_slots = []

    const slots = []
    const slotsBuffer = zymkey.getAllocSlotsList(false)
    for (let i = 0; i < slotsBuffer.length / 4; i++) {
      const slot = slotsBuffer.readInt32LE(4 * i)
      if (slot > 15) slots.push(slot)
    }

    for (const slot of slots) {
      const slotDetails = zymkey.getWalletNodeAddrFromKeySlot(slot)
      if (slotDetails.wallet_name == this.wallet_name) {
        if (slotDetails.node_address == ZymbitKeyring.base_path) {
          this.base_slot = slot
        } else if (slotDetails.node_address.includes(ZymbitKeyring.base_path + "/")) {
          slotDetails.slot = slot
          delete slotDetails.wallet_name
          this.account_slots.push(slotDetails)
        }
      }
    }
    if (this.base_slot == 0) {
      let deepestPath = { node_address: "m", slot: this.master_slot }
      for (const slot of slots) {
        const slotDetails = zymkey.getWalletNodeAddrFromKeySlot(slot)
        if (slotDetails.wallet_name == this.wallet_name) {
          if (ZymbitKeyring.base_path.includes(slotDetails.node_address) && slotDetails.node_address.length > deepestPath.node_address.length) {
            slotDetails.slot = slot
            delete slotDetails.wallet_name
            deepestPath = slotDetails
          }
        }
      }
      this.base_slot = this._generateBasePathKey(deepestPath)
    }
    return Promise.resolve()
  }


  addAccounts(n = 1) {
    const newAccounts = []
    if (n < 1) return newAccounts

    let nextAccountIndex = this.account_slots.reduce((prev, curr) => {
      if (Number(prev < Number(curr.node_address.slice(ZymbitKeyring.base_path.length + 1)))) {
        return Number(curr.node_address.slice(ZymbitKeyring.base_path.length + 1)) + 1
      } else return prev + 1
    }, 0)

    for (let i = 0; i < n; i++) {
      const slot = zymkey.genWalletChildKey(this.base_slot, nextAccountIndex, false, false)
      const slotDetails = zymkey.getWalletNodeAddrFromKeySlot(slot.slot)
      slotDetails.slot = slot.slot
      delete slotDetails.wallet_name
      this.account_slots.push(slotDetails)
      const newAccountEthAddress = ethers.computeAddress('0x' + bytesToHex(zymkey.exportPubKey(slotDetails.slot, false)))
      newAccounts.push(newAccountEthAddress)
      nextAccountIndex++
    }

    return Promise.resolve(newAccounts)
  }

  getAccounts() {
    return Promise.resolve(this.account_slots.map(obj => ethers.computeAddress('0x' + bytesToHex(zymkey.exportPubKey(obj.slot, false)))))
  }

  signTransaction(address, transaction) {
    const signingSlot = this._getSlot(address)
    if (!signingSlot) throw new Error("Keyring does not contain this address")

    const txHash = transaction.hash(true)
    const { signature, recovery_id } = zymkey.genECDSASigFromDigest(txHash, signingSlot)

    const validSignature = this._generateValidECDSASignature(signature, recovery_id, transaction.getChainId())
    transaction.r = validSignature.r
    transaction.s = validSignature.s
    transaction.v = validSignature.v

    return Promise.resolve(transaction)
  }

  signMessage(address, data) {
    const signingSlot = this._getSlot(address)
    if (!signingSlot) throw new Error("Keyring does not contain this address")

    const messageToSign = Buffer.from(stripHexPrefix(data), 'hex')
    const { signature, recovery_id } = zymkey.genECDSASigFromDigest(messageToSign, signingSlot)

    const validSignature = this._generateValidECDSASignature(signature, recovery_id)

    return Promise.resolve(concatSig(validSignature.v, validSignature.r, validSignature.s))
  }

  exportAccount(address) {
    throw new Error('Not supported on Zymbit devices')
  }

  removeAccount(address) {
    const slot = this._getSlot(address)
    if (!slot) throw new Error("Keyring does not contain this address")
    zymkey.removeKey(slot, false)
    for (let i = 0; i < this.account_slots.length; i++) {
      if(this.account_slots[i].slot == slot){
        delete this.account_slots[i]
      }
    }
    return Promise.resolve()
  }

  _generateBasePathKey(deepestPath) {
    let slot = 0
    switch (deepestPath.node_address) {
      case `m`:
        slot = zymkey.genWalletChildKey(deepestPath.slot, 44, true, false)
        break;
      case `m/44'`:
        slot = zymkey.genWalletChildKey(deepestPath.slot, 60, true, false)
        break;
      case `m/44'/60'`:
        slot = zymkey.genWalletChildKey(deepestPath.slot, 0, true, false)
        break;
      case `m/44'/60'/0'`:
        slot = zymkey.genWalletChildKey(deepestPath.slot, 0, false, false)
        break;
      case ZymbitKeyring.base_path:
        return deepestPath.slot
    }
    const slotDetails = zymkey.getWalletNodeAddrFromKeySlot(slot.slot)
    slotDetails.slot = slot.slot
    return this._generateBasePathKey(slotDetails)
  }

  _generateValidECDSASignature(signature, recovery_id, chain_Id = false) {
    const N = new BN('115792089237316195423570985008687907852837564279074904382605163141518161494337', 10)
    let s = new BN(signature.slice(32, 64).toString('hex'), 16)
    let y_parity = Boolean(recovery_id)
    if ((s.muln(2)).gte(N)) {
      y_parity = !y_parity
      s = N.sub(s)
    }

    const validSignature = {}
    validSignature.r = signature.slice(0, 32)
    validSignature.s = s.toBuffer()
    if (chain_Id) {
      validSignature.v = 35 + (chain_Id * 2) + (y_parity ? 1 : 0)
    } else {
      validSignature.v = 27 + (y_parity ? 1 : 0)
    }

    return validSignature
  }

  _getSlot(address) {
    try {
      return (this.account_slots.find(obj => ethers.computeAddress('0x' + bytesToHex(zymkey.exportPubKey(obj.slot, false))).toLowerCase() === address.toLowerCase())).slot
    } catch (e) {
      return false
    }
  }

}

ZymbitKeyring.type = type;
ZymbitKeyring.base_path = BIP44_ETH_BASE_PATH
module.exports = ZymbitKeyring;